frame/society/src/lib.rs

```diff
@@ -0,0 +1,1372 @@
+		/// A candidate was dropped (due to an excess of bids in the system).
+		AutoUnbid(AccountId),
+		/// A candidate was dropped (by their request).
+		Unbid(AccountId),
+		/// A candidate was dropped (by request of who vouched for them).
+		Unvouch(AccountId),
+		/// A group of candidates have been inducted. The batch's primary is the first value, the
+		/// batch in full is the second.
+		Inducted(AccountId, Vec<AccountId>),
+		/// A suspended member has been judged
+		SuspendedMemberJudgement(AccountId, bool),
+		/// A candidate has been suspended
+		CandidateSuspended(AccountId),
+		/// A member has been suspended
+		MemberSuspended(AccountId),
+		/// A member has been challenged
+		Challenged(AccountId),
+		/// A vote has been placed (candidate, voter, vote)
+		Vote(AccountId, AccountId, bool),
+		/// A vote has been placed for a defending member (voter, vote)
+		DefenderVote(AccountId, bool),
+	}
+}
+
+/// Pick an item at pseudo-random from the slice, given the `rng`. `None` iff the slice is empty.
+fn pick_item<'a, R: RngCore, T>(rng: &mut R, items: &'a [T]) -> Option<&'a T> {
+	if items.is_empty() {
+		None
+	} else {
+		Some(&items[pick_usize(rng, items.len() - 1)])
+	}
+}
+
+/// Pick a new PRN, in the range [0, `max`] (inclusive).
+fn pick_usize<'a, R: RngCore>(rng: &mut R, max: usize) -> usize {
+
+	(rng.next_u32() % (max as u32 + 1)) as usize
+}
+
+impl<T: Trait<I>, I: Instance> Module<T, I> {
+	/// Puts a bid into storage ordered by smallest to largest value.
+	/// Allows a maximum of 1000 bids in queue, removing largest value people first.
+	fn put_bid(who: T::AccountId, value: BalanceOf<T, I>, bid_kind: BidKind<T::AccountId, BalanceOf<T, I>>) {
+		<Bids<T, I>>::mutate(|bids| {
+			match bids.binary_search_by(|x| x.0.cmp(&value)) {
+				Ok(pos) | Err(pos) => bids.insert(pos, (value, who, bid_kind)),
+			}
+			// Keep it reasonably small.
+			if bids.len() > MAX_BID_COUNT {
+				let (_, popped, kind) = bids.pop().expect("b.len() > 1000; qed");
+				if let BidKind::Vouch(voucher, _) = kind {
+					<Vouching<T, I>>::remove(&voucher);
+				}
+				let _unreserved = T::Currency::unreserve(&popped, T::CandidateDeposit::get());
+				Self::deposit_event(RawEvent::AutoUnbid(popped));
+			}
+		});
+	}
+
+	/// Check a user is a bid.
+	fn is_bid(m: &T::AccountId) -> bool {
+		// Bids are ordered by `value`, so we cannot binary search for a user.
+		<Bids<T, I>>::get()
+			.iter()
+			.find(|x| x.1 == *m)
+			.is_some()
+	}
+
+	/// Check a user is a candidate.
+	fn is_candidate(m: &T::AccountId) -> bool {
+		// Candidates are ordered by `value`, so we cannot binary search for a user.
+		<Candidates<T, I>>::get()
+			.iter()
+			.find(|x| x.1 == *m)
+			.is_some()
+	}
+
+	/// Check a user is a member.
+	fn is_member(m: &T::AccountId) -> bool {
+		<Members<T, I>>::get()
+			.binary_search(m)
+			.is_ok()
+	}
+
+	/// Add a member to the sorted members list. Will not add a duplicate member.
+	fn add_member(m: &T::AccountId) -> DispatchResult {
```

Comment:
This is a dangerous function as it can both return an `Err` and has global side-effects.

Docs should state as much and make it clear that it *must* be at the border point between checks and actions in a dispatch function.

Reply:
Made it so that this function cannot fail. If the user is already a member, we simply do nothing :)